File: CSES_problems/multiplication_table.py

n = int(input())
obj = (n**2)//2
l, r = 1, n**2
# l: la més petita trobada que funciona
# r: la més petita trobada que no funciona
while r-l > 1:
    m = (r+l)//2
    trobat = 0
    less = 0
    for i in range(1, n+1):
        less += min(m//i, n)
        if m%i == 0 and m//i <= n:
            trobat += 1
    if less-trobat > obj:
        r = m
    elif less <= obj:
        l = m+1
    else:
        l = m
        break
print(l)

File: CSES_problems/multiplication_table.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
    int n;
    cin >> n;
    long long obj = (1LL * n * n) / 2;
    long long l = 1, r = 1LL * n * n;

    while (r - l > 1) {
        long long m = (r + l) / 2;
        long long less = 0, trobat = 0;

        for (int i = 1; i <= n; ++i) {
            less += min(m / i, (long long)n);
            if (m % i == 0 && m / i <= n) {
                ++trobat;
            }
        }

        if (less - trobat > obj) {
            r = m;
        } else if (less <= obj) {
            l = m + 1;
        } else {
            l = m;
            break;
        }

    }

    cout << l << endl;
    return 0;
}
